### DISCUSSION TITLE
Attempt to access prevValue, newValue... fires twice

### DISCUSSION THREAD

**Discussion by chrissysemens** [pmndrs/zustand#1838] [Bug report] [answered]
Hi,

I'm trying to use the prevState, newState callback which is used to add a custom compare, to carry out an additional action.
I'm trying avoid using an additional useEffect, to watch for changes on the data.

I can see the values I want but it seems to be firing twice.. is this expected behaviour? 
Is there a better way to access prevState in the consuming component?

![image](https://github.com/pmndrs/zustand/assets/46443589/dbbf5353-8ee3-4508-98cd-3fc72071e824)

![image](https://github.com/pmndrs/zustand/assets/46443589/bc7c42e0-216a-43d1-93f6-2973eaa6246c)

**Comment by dai-shi** [comment]
Can you create a smallest possible reproduction with https://codesandbox.io/s/react-new ?

**Comment by chrissysemens** [comment]
Yeah, sure mate! I'll post it tomorrow if its ok.

**Comment by chrissysemens** [comment]
Hi @dai-shi 
https://codesandbox.io/s/example-qxdjqd

I was just looking to see if there was a way of accessing prevValue, newValue on calling an update function. I know this isn't the way I should be doing it. Is there any way, to do that?

**Reply by dai-shi** [reply, marked as answer]
Ah, I see.
If you want to avoid `useEffect`, you want to use `.subscribe`.

https://codesandbox.io/s/example-forked-qcis09?file=/src/App.tsx

**Reply by chrissysemens** [reply]
Thanks so much! Thats amazing!
